### PR TITLE
Complete python-jose typing

### PR DESCRIPTION
**Pull request by hasier**
Complete the work started in https://github.com/python/typeshed/pull/8238 and make minor fixes to changes included there.

### CODE REVIEW

File: stubs/python-jose/jose/jws.pyi [hunk 1]
```diff
@@ -4,21 +4,21 @@ from typing import Any
 from .backends.base import Key
 
 def sign(
-    payload: str | Mapping[str, Any],
+    payload: bytes | Mapping[str, Any],
```
Comment (hasier) [inline stubs/python-jose/jose/jws.pyi:7]: It is passed down to `base64url_encode`, which expects `bytes`.
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L42
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L153

File: stubs/python-jose/jose/jws.pyi [hunk 2]
```diff
@@ -4,21 +4,21 @@ from typing import Any
...
     # Internally it's passed down to jwk.construct(), which explicitly checks for
     # key as dict instance, instead of a Mapping
-    key: str | dict[str, Any] | Key,
+    key: str | bytes | dict[str, Any] | Key,
```
Comment (hasier) [inline stubs/python-jose/jose/jws.pyi:10]: It is passed down to a `Key` class after calling `jwk.construct()`, which accounts for both `str` and `bytes`. Example https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/backends/cryptography_backend.py#L78-L94

File: stubs/python-jose/jose/jws.pyi [hunk 3]
```diff
@@ -4,21 +4,21 @@ from typing import Any
...
     headers: Mapping[str, Any] | None = ...,
     algorithm: str = ...,
 ) -> str: ...
 def verify(
-    token: str,
-    key: str | Mapping[str, Any] | Key,
+    token: str | bytes,
```
Comment (hasier) [inline stubs/python-jose/jose/jws.pyi:15]: It is passed down to `_load`, which accounts for both `str` and `bytes`.
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L70
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L173-L174

File: stubs/python-jose/jose/jws.pyi [hunk 4]
```diff
@@ -4,21 +4,21 @@ from typing import Any
...
+    key: str | bytes | Mapping[str, Any] | Key,
```
Comment (hasier) [inline stubs/python-jose/jose/jws.pyi:16]: Same as https://github.com/python/typeshed/pull/8258/files#r916985268

File: stubs/python-jose/jose/jws.pyi [hunk 5]
```diff
@@ -4,21 +4,21 @@ from typing import Any
...
     # Callers of this function, like jwt.decode(), and functions called internally,
     # like jws._verify_signature(), use and accept algorithms=None
     algorithms: str | Container[str] | None,
     verify: bool = ...,
-) -> str: ...
+) -> bytes: ...
```
Comment (hasier) [inline stubs/python-jose/jose/jws.pyi:21]: It is returned from `base64url_decode`, which returns `bytes`.
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L70
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L193
- https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/utils.py#L66

File: stubs/python-jose/jose/jwt.pyi [hunk 6]
```diff
@@ -6,14 +6,14 @@ from .backends.base import Key
 def encode(
     claims: MutableMapping[str, Any],
     # Internally it calls jws.sign() that expects a key dict instance instead of Mapping
-    key: str | dict[str, Any] | Key,
+    key: str | bytes | dict[str, Any] | Key,
```
Comment (hasier) [inline stubs/python-jose/jose/jwt.pyi:9]: Same as https://github.com/python/typeshed/pull/8258/files#r916985268

File: stubs/python-jose/jose/jwt.pyi [hunk 7]
```diff
@@ -6,14 +6,14 @@ from .backends.base import Key
...
     algorithm: str = ...,
     headers: Mapping[str, Any] | None = ...,
     access_token: str | None = ...,
 ) -> str: ...
 def decode(
-    token: str,
-    key: str | Mapping[str, Any] | Key,
+    token: str | bytes,
```
Comment (hasier) [inline stubs/python-jose/jose/jwt.pyi:15]: It is passed down to `jws.verify()`, which accepts both `str` and `bytes` https://github.com/python/typeshed/pull/8258/files#r916986297

File: stubs/python-jose/jose/utils.pyi [hunk 8]
```diff
@@ -1,9 +1,15 @@
...
-def ensure_binary(s): ...
+from collections.abc import Iterable
+from datetime import timedelta
+from hashlib import _Hash
+from typing import Any, Callable
+
+
+def long_to_bytes(n: int, blocksize: int | None = ...) -> None: ...
+def long_to_base64(data: int, size: int | None = ...) -> bytes: ...
+def int_arr_to_long(arr: Iterable[Any]) -> int: ...
```
Comment (hasier) [inline stubs/python-jose/jose/utils.pyi:8]: Not too sure about this one tbh 😕  Happy to remove it if it's not helpful.

**Review by hauntsaninja — COMMENTED** [review]
If this actually adds all useful missing definitions to the stub, consider adding
```
[tool.stubtest]
ignore_missing_stub = false
```
to METADATA.toml

This will make sure that stubtest complains if any future additions are left out of the stub

**Comment by hasier** [comment]
> If this actually adds all useful missing definitions to the stub, consider adding
> 
> ```
> [tool.stubtest]
> ignore_missing_stub = false
> ```
> 
> to METADATA.toml
> 
> This will make sure that stubtest complains if any future additions are left out of the stub

Thanks for that @hauntsaninja! Will do if there's positive feedback about https://github.com/python/typeshed/pull/8258#discussion_r916989844, otherwise I guess we don't want that flag in place?

**Comment by hauntsaninja** [comment]
If it is decided those are useless and should be removed, we could add them to an allowlist. This will tell stubtest to ignore those, but ensure that everything else is complete. For an example, see https://github.com/python/typeshed/blob/b145b32aa988bcb0c85ece32ce64df521fcafafb/stubs/PyMySQL/%40tests/stubtest_allowlist.txt#L1

**Comment by hasier** [comment]
Any idea on what this error is about? https://github.com/python/typeshed/runs/7257227823?check_suite_focus=true
I guess it has to do with these imports? If so, any suggestions on how to type them, given that it's possible they may not exist? https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwk.py#L5-L28

**Comment by hauntsaninja** [comment]
Not too familiar with python-jose, so take with a grain of salt. It looks like these imports are part of the public API of `jose.jwk` so they should ideally be in the stub. This then looks to be a very similar situation to `jose.backends.__init__`.

https://github.com/python/typeshed/blob/72c7eb6f80471b70c30a9e871c8d950082b6874f/stubs/python-jose/jose/backends/__init__.pyi#L14-L21

So we just add all of them in the stub (and trust the user to know whether or not they're actually available). And if stubtest complains that those imports don't actually exist in the CI environment, we just silence it here: https://github.com/python/typeshed/blob/master/stubs/python-jose/%40tests/stubtest_allowlist.txt

**Comment by hasier** [comment]
@hauntsaninja fixed all the CI issues and it seems to be working now. I'd appreciate some 👀 to get it merged, not sure if you can help me with this. CC @AlexWaygood as you helped me with [the previous PR](https://github.com/python/typeshed/pull/8238) and got it merged 🙂

File: stubs/python-jose/jose/jwe.pyi [hunk 9]
```diff
@@ -1,13 +1,23 @@
...
-    kid: Any | None = ...,
-): ...
-def decrypt(jwe_str: str, key: str | dict[str, str]): ...
-def get_unverified_header(jwe_str: str): ...
+    plaintext: str | bytes,
+    # Internally it's passed down to jwk.construct(), which explicitly checks for
+    # key as dict instance, instead of a Mapping
+    key: str | bytes | dict[str, Any] | Key,
+    encryption: str | None = ...,
+    algorithm: str | None = ...,
```
[the remark this answers is not in the document]
Comment (hasier) [reply stubs/python-jose/jose/jwe.pyi]: True, done! https://github.com/python/typeshed/pull/8258/commits/53413876fe980ab99278b820878dac13e3a91927

File: stubs/python-jose/jose/jwe.pyi [hunk 10]
```diff
@@ -1,13 +1,23 @@
...
+    cty: str | None = ...,
+    kid: str | None = ...,
+) -> bytes: ...
+def decrypt(
+    jwe_str: str | bytes,
+    # Internally it's passed down to jwk.construct(), which explicitly checks for
+    # key as dict instance, instead of a Mapping
+    key: str | bytes | dict[str, Any] | Key,
+) -> bytes | None: ...
+def get_unverified_header(jwe_str: str | None) -> Mapping[str, Any]: ...
```
[the remark this answers is not in the document]
Comment (hasier) [reply stubs/python-jose/jose/jwe.pyi]: Ah nice, thanks for explaining. Done! https://github.com/python/typeshed/pull/8258/commits/53413876fe980ab99278b820878dac13e3a91927

File: stubs/python-jose/jose/jwk.pyi [hunk 11]
```diff
@@ -1,7 +1,12 @@
...
 
-from .backends.base import Key as Key
+from .backends import AESKey as AESKey, ECKey as ECKey, HMACKey as HMACKey, RSAKey as RSAKey
+from .backends.base import DIRKey as DIRKey, Key
 
-def get_key(algorithm): ...
-def register_key(algorithm, key_class: Key): ...
-def construct(key_data, algorithm: Any | None = ...): ...
+def get_key(algorithm: str) -> type[Key] | None: ...
+def register_key(algorithm: str, key_class: type[Key]) -> bool: ...
```
[the remark this answers is not in the document]
Comment (hasier) [reply stubs/python-jose/jose/jwk.pyi]: Done! https://github.com/python/typeshed/pull/8258/commits/53413876fe980ab99278b820878dac13e3a91927

File: stubs/python-jose/jose/utils.pyi [hunk 12]
```diff
@@ -1,9 +1,14 @@
...
-def base64url_decode(input): ...
-def base64url_encode(input): ...
-def timedelta_total_seconds(delta): ...
-def ensure_binary(s): ...
+from collections.abc import Callable, Iterable
+from datetime import timedelta
+from hashlib import _Hash
+from typing import Any
+
+def long_to_bytes(n: int, blocksize: int | None = ...) -> None: ...
```
[the remark this answers is not in the document]
Comment (hasier) [reply stubs/python-jose/jose/utils.pyi]: 🤦 Done! https://github.com/python/typeshed/pull/8258/commits/53413876fe980ab99278b820878dac13e3a91927

**Review by AlexWaygood — APPROVED** [review]
Thanks for all the work done here! Sorry we were a few days in getting to this.

**Pull request merged**